cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -716,17 +738,31 @@ class ObjectAppendStream final : public io::OutputStream {
     io::internal::CloseFromDestructor(this);
   }
 
-  Status Init() {
-    if (content_length_ != kNoSize) {
-      DCHECK_GE(content_length_, 0);
-      pos_ = content_length_;
+  Status Init(const bool truncate,
+              std::function<Status()> ensure_not_flat_namespace_directory) {
```

Comment:
You can inject `AzureFileSystem *azure_file_system` here and not have to allocate a closure for this. You would call  `AzureFileSystem::Impl::EnsureNotFlatNamespaceDirectory(location)` via `azure_file_system->impl_` (accessible because the handles produced by the azure file system can be friends with the filesystem class).

Reply:
The trick is to forward-declare the class as well.

```cpp
diff --git a/cpp/src/arrow/filesystem/azurefs.h b/cpp/src/arrow/filesystem/azurefs.h
index 2a131e40c..d48ef9dd7 100644
--- a/cpp/src/arrow/filesystem/azurefs.h
+++ b/cpp/src/arrow/filesystem/azurefs.h
@@ -44,6 +44,7 @@ class DataLakeServiceClient;

 namespace arrow::fs {

+class ObjectAppendStream;
 class TestAzureFileSystem;

 /// Options for the AzureFileSystem implementation.
@@ -180,6 +181,7 @@ class ARROW_EXPORT AzureFileSystem : public FileSystem {

   explicit AzureFileSystem(std::unique_ptr<Impl>&& impl);

+  friend class ObjectAppendStream;
   friend class TestAzureFileSystem;
   void ForceCachedHierarchicalNamespaceSupport(int hns_support);
```

Reply:
I think my main problem was that `ObjectAppendStream` is defined inside an anonymous namespace. 

Are you suggesting to use `AzureFileSystem *azure_file_system` or `AzureFileSystem:Impl *azure_file_system` as the argument to `ObjectAppendStream::Impl`. I don't know how I can get a `AzureFileSystem` pointer from inside `AzureFileSystem::Impl` and using `AzureFileSystem::Impl` as the argument leads to `incomplete type` errors which I don't think I can avoid. 

Also if you wouldn't mind I would be interested to know what the disadvantage of a lambda function is compared to what you proposed. 

Sorry about my lacking C++ knowledge here.

Reply:
> Also if you wouldn't mind I would be interested to know what the disadvantage of a lambda function is compared to what you proposed.

To create the `std::function`, you heap allocate an object with copies of the values in the capture list and generate a lot more extra code in the binary:

```
class function {
  T valuesfromthecpapturelist;
  
  RetType operator()(ArgsType ...) {...};
}
```

When you think about an `std::function` this way (a pair of context data and a function), you realize the class you already serves that purpose.

But hey, this is becoming challenging, so I won't hold the PR anymore because of this. Moving to `Init()` was a big step in the right direction.

Reply:
Thanks for explaining